### PR TITLE
Make index creation agnostic of filesystem using fsspec

### PR DESCRIPTION
**Pull request by rom1504**

### CODE REVIEW

**Comment by rom1504** [comment]
Can you update the doc with the new params ?

**Comment by rom1504** [comment]
And I think we can delete the download utility and related files now

**Comment by rom1504** [comment]
It seems the CI is failing

File: autofaiss/datasets/readers/embeddings_iterators.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+    i = 0
+    with ThreadPool(50) as p:
+        for c in p.imap_unordered(lambda f: get_number_of_lines(f, fs), files):
+            count += c
+            i += 1
+
+    return count, dim
+
+
+def convert_parquet_to_numpy(
```
Comment (rom1504) [inline autofaiss/datasets/readers/embeddings_iterators.py:108]: I think this function can be removed

File: docs/getting_started/quantization.rst [hunk 2]
```diff
@@ -23,6 +23,34 @@ The ``autofaiss quantize`` command takes the following parameters:
 |                            |          | there are several files,   |
 |                            |          | they are read in the       |
 |                            |          | lexicographical order.     |
+|                            |          | This can be a local path   |
```
Comment (rom1504) [inline docs/getting_started/quantization.rst:26]: Can you do the same update in the readme ?

**Comment by rom1504** [comment]
pushed 36774aa3db1280c738607cff5c7e19bc9a670276 so numpy file reading stays fast (it was fast before thanks to memmap but we removed it due to fsspec)

File: docs/getting_started/quantization.rst (hunk 2 above)
Comment (rom1504) [reply docs/getting_started/quantization.rst:26]: done

**Comment by rom1504** [comment]
tested to work the same as before with local numpy

**Comment by rom1504** [comment]
Note : to avoid reading the embeddings parquet a second time, we could consider extracting, yielding and saving the keys from the parquet files in the read embeddings function.
These keys could be saved either as parquet, either in some format convenient for fast random access (eg arrow, hdf5 for one way, leveldb for 2 way).
That would probably be convenient but let's keep this for another PR

File: autofaiss/external/build.py [hunk 3]
```diff
@@ -189,7 +136,15 @@ def build_index(
             )
 
             # Extract training vectors
-            train_vectors = next(read_embeddings_local(embeddings_path, batch_size=train_size, verbose=True))
+            train_vectors = next(
```
Comment (rom1504) [inline autofaiss/external/build.py:139]: need to check that we are releasing the ram of that file here, I think it's not the case due to the fact it's an iterator

Comment (rom1504) [reply autofaiss/external/build.py:139]: maybe the fact we lose the reference to the iterator triggers the release

Comment (Mbompr) [reply autofaiss/external/build.py:139]: Calling next again deletes the previous one

```python
from dataclasses import dataclass

@dataclass
class WithDestructor(object):
    i: int
    def __del__(self):
        print(f"delete {self.i}")


def my_destructors(n):
    for i in range(n):
        yield WithDestructor(i)


n = 3
i = 0
destructors_iter = my_destructors(n)

while i < n:
    print(f'i={i}')
    a = next(destructors_iter)
    print(a)
    i += 1
````

Outputs
```
i=0
WithDestructor(i=0)
i=1
delete 0
WithDestructor(i=1)
i=2
delete 1
WithDestructor(i=2)
delete 2
```

Comment (rom1504) [reply autofaiss/external/build.py:139]: ok sounds good

**Pull request merged**